### PR TITLE
Fix handling of time exceeded exception in fetch phase

### PR DESCRIPTION
**Pull request by javanna**
The fetch phase is subject to timeouts like any other search phase. Timeouts may happen when low level cancellation is enabled (true by default), hence the directory reader is wrapped into ExitableDirectoryReader and a timeout is provided to the search request.

The exception that is used is TimeExceededException, but it is an internal exception that should never be returned to the user. When that is thrown, we need to catch it and throw error or mark the response as timed out depending on whether partial results are allowed or not.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/fetch/FetchPhaseDocsIterator.java [hunk 1]
```diff
@@ -71,17 +79,28 @@ public final SearchHit[] iterate(SearchShardTarget shardTarget, IndexReader inde
...
+                throw new SearchTimeoutException(shardTarget, "Time exceeded");
             }
+        } catch (Exception e) {
+            purgeSearcHits(searchHits);
             throw new FetchPhaseExecutionException(shardTarget, "Error running fetch phase for doc [" + currentDoc + "]", e);
         }
         return searchHits;
     }
 
+    private static void purgeSearcHits(SearchHit[] searchHits) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/search/fetch/FetchPhaseDocsIterator.java:96]: Typo :)

File: server/src/test/java/org/elasticsearch/action/search/FetchSearchPhaseTests.java [hunk 2]
```diff
@@ -749,4 +779,150 @@ private static void addProfiling(boolean profiled, QuerySearchResult queryResult
...
     }
+
+    public void testFetchTimeoutWithPartialResults() throws IOException {
+        Directory dir = newDirectory();
+        RandomIndexWriter w = new RandomIndexWriter(random(), dir);
+        w.addDocument(new Document());
+        IndexReader r = w.getReader();
+        w.close();
+        try {
+            ContextIndexSearcher contextIndexSearcher = new ContextIndexSearcher(r, null, null, new QueryCachingPolicy() {
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/action/search/FetchSearchPhaseTests.java:790]: NIT: maybe dry up the identical `ContextIndexSearcher` overrides :) ?

**Review by original-brownbear — APPROVED** [review]
LGTM

File: server/src/test/java/org/elasticsearch/action/search/FetchSearchPhaseTests.java (hunk 2 above)
Comment (javanna) [reply server/src/test/java/org/elasticsearch/action/search/FetchSearchPhaseTests.java]: done!

File: server/src/main/java/org/elasticsearch/search/fetch/FetchPhaseDocsIterator.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/fetch/FetchPhaseDocsIterator.java]: thanks!

**Pull request merged**